### PR TITLE
toolchain3

### PR DESCRIPTION
**Pull request by isuruf**
This uses the compiler wrappers implemented in spack package manager and creates a new toolchain package. Link [here](https://github.com/LLNL/spack/blob/master/lib/spack/env/cc). Macports has a similar idea in their tracker, https://trac.macports.org/ticket/44413

There is a bash script `cc` installed in `$PREFIX/bin/cf/` that looks at the arguments passed to it and process them before sending it to the real `cc` compiler. There are symlinks to `gcc, clang, g++, gfortran` etc.

This script looks at what mode the compiler/linker is invoked in and sets the mode. 'mode' is set to one of:
```
#    vcheck  version check
#    cpp     preprocess
#    cc      compile
#    as      assemble
#    ld      link
#    ccld    compile & link
```
1. `CONDA_FORGE_CFLAGS` is sent to the compiler if `cc` and `ccld`.
2. `CONDA_FORGE_CPPFLAGS` which includes `-I${PREFIX}/include` is sent in `cc, ccld, as, cpp`. 
3. `CONDA_FORGE_LDFLAGS` which includes `-L${PREFIX}/lib` is sent in `ld, ccld`.
4.  `rpath` is sent in `ccld` and `ld` modes with `-Wl,-rpath` and `-rpath`.

This avoids issues we've seen in toolchain package.

1. numpy.distutils packages like scipy.
Setting `LDFLAGS` env variable overwrites the `LDFLAGS` variable in numpy.distutils instead of appending. In toolchain3 we are not setting `LDFLAGS` at all, but sending them directly to the compiler and linker.

2. rpy2
rpy2's build system checks `CFLAGS` and preprocess them and chokes when `-headerpad_max_install_names` is set. toolchain3 does not have this problem.

This package also supports setting `CFLAGS` and `CXXFLAGS` which are sent to the compiler through the build system.

@conda-forge/core , what do you think?

### CODE REVIEW

**Comment by isuruf** [comment]
This also uses `ccache` if it is found in the build environment. cc @gqmelo

**Comment by isuruf** [comment]
This is ready now. I've added scipy and rpy2 as examples. Will remove if this is going to get merged.

**Comment by isuruf** [comment]
Ping

File: recipes/toolchain3/build.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# Copy the [de]activate scripts to $PREFIX/etc/conda/[de]activate.d.
+# This will allow them to be run on environment activation.
+for CHANGE in "activate" "deactivate"
+do
+    mkdir -p "${PREFIX}/etc/conda/${CHANGE}.d"
+    cp "${RECIPE_DIR}/${CHANGE}.sh" "${PREFIX}/etc/conda/${CHANGE}.d/toolchain3_${CHANGE}.sh"
+done
+
+mkdir -p ${PREFIX}/bin/cf
+cd ${PREFIX}/bin/cf
```
Comment (scopatz) [inline recipes/toolchain3/build.sh]: Is it probably better to spell out the word `conda-forge` here

Comment (isuruf) [inline recipes/toolchain3/build.sh]: I'll make the change

File: recipes/toolchain3/conda-forge-cc [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+##############################################################################
+# Copyright (c) 2013-2016, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
```
Comment (scopatz) [inline recipes/toolchain3/conda-forge-cc]: Is there any way to remove and reimplement the LGPL parts of this?

Comment (isuruf) [inline recipes/toolchain3/conda-forge-cc]: This LICENSE was for the whole spack codebase. I can ask the spack devs whether they'd like to re-license just this file if there's a need.

Comment (scopatz) [inline recipes/toolchain3/conda-forge-cc]: Does this whole file come from spack?

Comment (isuruf) [inline recipes/toolchain3/conda-forge-cc]: Pretty much. I removed some unnecessary stuff. https://github.com/LLNL/spack/blob/master/lib/spack/env/cc

**Comment by scopatz** [comment]
Hi @isuruf, thanks for putting this in. From the perspective that this is just another package, I see no reason to not merge this in (after scipy and rpy2 are removed). People would be free to use this or not use it as they see want.

The broader and harder question is about whether is this supposed to replace the existing toolchain package. If so, why not have this as a PR to the toolchain repo? Maybe I am missing something but it doesn't seem like the example recipes had to change at all.

**Comment by isuruf** [comment]
Thanks for the review.  In theory, this should be compatible with the existing toolchain, but I'm not sure what it'll break. Hence a new package. Also, I wanted to test out packages that had issues with `toolchain` first. This way it's easier to test them out. 

scipy's build script [here](https://github.com/conda-forge/scipy-feedstock/blob/master/recipe/build.sh) is just 2 lines with this package. Previous build script also works.

I'm open to either of the 2 options. New package or as a PR to toolchain. Let me know what I should do.

**Comment by scopatz** [comment]
It is probably good to have other members of @conda-forge/core comment, but I think that it is probably best to have a single toolchain package, rather than make the situation more complex. As such a PR into the toolchain feedstock would be my preference.

**Comment by isuruf** [comment]
I've made a PR here, https://github.com/conda-forge/toolchain-feedstock/pull/23

**Comment by ocefpaf** [comment]
@isuruf I don't really use any `toolchain` that set flags in my packages, nor I think we should enforce any flags beyond those required by the compilers we ship (I won't elaborate on this for brevity).

With that said I do understand that maintaining a huge number of packages that require some common flags, like what you are doing with the sage ecosystem, requires a package like this.

I don't really know the best approach but I would like to suggest a quick solution for your case. Why not a package named `sage-toolchain` that you can use without worrying too much the need of all of conda-forge's ecosystems? With time we could keep discussing the functionality you have there and migrate them to a more general `toolchain`, but it seems that you have a need for it now and this discussion is stalled at the moment.

**Comment by scopatz** [comment]
Maybe we should just merge this then? I don't mind merging this. I originally thought it would be cleaner to have one package, but that didn't quite work out.

**Comment by ocefpaf** [comment]
What do you say @isuruf?

**Comment by isuruf** [comment]
It'll be great to have this in and I can improve this incrementally then.

**Comment by scopatz** [comment]
OK then let's merge then!

**Comment by isuruf** [comment]
Thanks

**Comment by jakirkham** [comment]
> It'll be great to have this in and I can improve this incrementally then.

Ah. Thought this was getting folded into the `ccache-toolchain` for some reason. ( https://github.com/conda-forge/staged-recipes/pull/3135 )

**Comment by isuruf** [comment]
@jakirkham, that's not a bad idea either. Since we want to do PRs with `ccache-toolchain` we can have it depend on `toolchain3` instead of `toolchain`

**Comment by scopatz** [comment]
@jakirkham I don't think so, it is more of a replacement than an addition.  Making it an addition would probably require more workarounds, right?

**Comment by isuruf** [comment]
> Making it an addition would probably require more workarounds, right?

Yes.

**Pull request merged**